### PR TITLE
1x1 convolution acceleration

### PR DESCRIPTION
**Pull request by chinakook**
* GEMM directly without im2col or col2im in 1x1 convolution(stride=1,pad=0). The 1x1 convolution is used very common in modern CNN networks such as Googlenet/Inception/Resnet/Mobilenet etc.

### CODE REVIEW

File: src/operator/convolution-inl.h [hunk 1]
```diff
@@ -212,12 +216,41 @@ class ConvolutionOp : public Operator {
...
     Tensor<xpu, 4, DType> out_grad_4d = out_grad[conv::kOut].get_with_shape<xpu, 4, DType>(
       Shape4(num_, group_, K, N), s);
-    Tensor<xpu, 3, DType> col_buffer_3d = col_buffer.get_with_shape<xpu, 3, DType>(
-      Shape3(group_, M, N), s);
     // For computing dLoss/dWeight
     Tensor<xpu, 3, DType> dweight_3d = in_grad[conv::kWeight].get_with_shape<xpu, 3, DType>(
       Shape3(group_, K, M), s);
 
+  // no need to allocating memory and reordering in memory
+  if (is_1x1_) {
```
Comment (reminisce) [inline src/operator/convolution-inl.h]: Is indentation correct here?

**Comment by chinakook** [comment]
@reminisce I will complete the testings. I think performance improvement and memory reduction will show up in the image detection and segmentation cases where feature maps are very big. You can refer to the tensorflow's optimization for the 1x1 conv, it's more simple and more clear than Caffe:
[tensorflow conv forward](https://github.com/tensorflow/tensorflow/blob/a0d784bdd31b27e013a7eac58a86ba62e86db299/tensorflow/core/kernels/conv_ops_using_gemm.cc#L238)
[tensorflow conv backward](https://github.com/tensorflow/tensorflow/blob/42ca99b5aae03a8122ba0db94abfe1f3f5c257dc/tensorflow/core/kernels/conv_grad_input_ops.cc#L666)

**Comment by chinakook** [comment]
Performance test script:
import mxnet as mx

def benchmark_convolution(pass_type="whole", device_type="cpu", input_shape=None, kernel=None, stride=None, pad=None, cudnn_off=True):
    if input_shape is None:
        raise RuntimeError("Input shape must be (N, C, W) or (N, C, H, W) or (N, C, D, H, W")
    if kernel is None:
        raise RuntimeError("Kernel cannot be None and must match with dim of images")
    data = mx.sym.Variable(name="data")
    num_filter = 3
    if device_type == "cpu":
        dev = mx.cpu()
    else:
        dev = mx.gpu()
    conv = mx.sym.Convolution(data=data, kernel=kernel, stride=stride, pad=pad, num_filter=num_filter, no_bias="true",  cudnn_off=cudnn_off, name="conv")
    arg_shapes, out_shapes, _ = conv.infer_shape(data=input_shape)
    input_data = mx.random.uniform(-5, 5, input_shape, ctx=dev)
    out_grad = input_data
    args = {}
    args["data"] = input_data
    args["conv_weight"] = mx.random.normal(0, 1, (num_filter, input_shape[1]) + kernel, ctx=dev)
    args_grad = [mx.nd.empty(s) for s in arg_shapes]
    return mx.test_utils.check_speed(sym=conv, N=10, ctx=dev, typ=pass_type, data=input_shape, args=args, args_grad=args_grad)

if __name__ == '__main__':
    print "v2, cpu, 1D: ", benchmark_convolution(input_shape=(10, 16, 512), kernel=(1,), stride=(1,), pad=(0,))
    print "v2, gpu, 1D: ", benchmark_convolution(device_type="gpu", input_shape=(10, 16, 512), kernel=(1,), stride=(1,), pad=(0,))
    print "\n"
    print "v2, cpu, 2D: ", benchmark_convolution(input_shape=(10, 16, 64, 64), kernel=(1, 1), stride=(1,1), pad=(0,0))
    print "v2, gpu, 2D: ", benchmark_convolution(device_type="gpu", input_shape=(10, 16, 64, 64), kernel=(1, 1), stride=(1,1), pad=(0,0))
    print "cudnn, gpu, 2D: ", benchmark_convolution(device_type="gpu", cudnn_off=False, input_shape=(10, 16, 64, 64), kernel=(1, 1), stride=(1,1), pad=(0,0))
    print "\n"
    print "v2, cpu, 3D: ", benchmark_convolution(input_shape=(4, 8, 10, 30, 20), kernel=(1, 1, 1), stride=(1,1,1), pad=(0,0,0))
    print "v2, gpu, 3D: ", benchmark_convolution(device_type="gpu", input_shape=(4, 8, 10, 30, 20), kernel=(1, 1, 1), stride=(1,1,1), pad=(0,0,0))
    print "cudnn, gpu, 3D: ", benchmark_convolution(device_type="gpu", cudnn_off=False, input_shape=(4, 8, 10, 30, 20), kernel=(1, 1, 1), stride=(1,1,1), pad=(0,0,0))

1x1 convolution unoptimized version:
v2, cpu, 1D:  0.00665521621704
v2, gpu, 1D:  0.000849199295044

v2, cpu, 2D:  0.0047404050827
v2, gpu, 2D:  0.00151178836823
cudnn, gpu, 2D:  0.000375509262085

v2, cpu, 3D:  0.0116127967834
v2, gpu, 3D:  0.00107021331787
cudnn, gpu, 3D:  0.0016508102417

1x1 convolution optimized version:
v2, cpu, 1D:  0.000251388549805
v2, gpu, 1D:  0.000379395484924

v2, cpu, 2D:  0.00151929855347
v2, gpu, 2D:  0.001118683815
cudnn, gpu, 2D:  0.000412487983704

v2, cpu, 3D:  0.00025269985199
v2, gpu, 3D:  0.000564694404602
cudnn, gpu, 3D:  0.00160720348358

**Comment by piiswrong** [comment]
please rebase to master

**Comment by chinakook** [comment]
Is that rebased? I found It's a little confused.

**Comment by reminisce** [comment]
It seems not rebased. The **File Changed** tab should only contain your changes.
Suppose `upstream` is the alias to https://github.com/apache/incubator-mxnet in your local git. Then type
`git pull --rebase upstream master` to sync with the latest code base and put your commits on top the git history. Type `git log`, you should see your commits for this PR sit on top of the latest upstream commit history.

**Comment by piiswrong** [comment]
use
```
git fetch upstream
git rebase upstream/master
```
to rebase to master.

**Pull request merged**